### ISSUE TITLE
Getting eigenvalues for ILR ordination

### ISSUE THREAD

**Issue by davidwrandall** [reptalex/phylofactor#34]
Hi, thanks so much for creating PhyloFactor - it's a great package which has helped enormously with my data analysis!

I wonder if I could ask whether it is possible to retrieve the Eigenvalues for ILR ordinations created using the pf.ILRprojection command? Eg to create a scree plot or add to the axes of an ordination plot?

Many thanks,

David

**Comment by reptalex** [comment]
Hi David!

Glad you’re enjoying the package!

A few semantic things - I don’t believe there are any eigenvalues involved in ILR ordination - rather, this is just a projection of the data onto a low-dimensional subspace spanned by the contrast vectors in `pf$basis`. Hence: you have the vectors in the basis and their projections, and the ILR ordination is produced by

`t(pf$basis[,factors]) %*% log(pf$Data)`

Hope that helps!

**Comment by davidwrandall** [comment]
Okay - thanks - that's very helpful. And thanks for the very speedy response.


David
________________________________

**Comment by davidwrandall** [comment]
Hi Alex,

One other question (I hope you don't mind) - does the PhyloFactor package have any built-in multiple hypothesis testing? I'm trying to understand whether or not this would be appropriate, and how it might work in practice, in the context of phylogenetic factorisation?

With much thanks,


David
________________________________

**Comment by reptalex** [comment]
Hey David!

TLDR: no. 

PhyloFactor does, however, report the uncorrected P-values. If you have stop.early=T as your input, then at every iteration there will be a Kolmogorov-Smirnov test of the uniformity of those P-values and, if they aren't different from uniform (in a one-tailed test looking for hockey-sticked P-value distributions), the algorithm will stop. In the original PeerJ article and I believe again in the Ecological Monograph, I showed that this works reasonably well at ensuring a conservative estimate for the number of significant factors (I compared it with the null simulation approach - Horn's parallel analysis - and show in silico that this KS-based stopping criterion is suitable and far less computationally intensive).  If the KS-test rejects the uniform P-value hypothesis, then a the new factor is added (drawing out the edge with the highest objective statistic i.e. minimum P-value).

Beyond that, there's some really interesting features about the sequence of P-values you obtain from phylofactorization. They don't correspond exactly to the order statistics of a uniform random variable (minimum, second smallest, third smallest etc.) but they do correspond neatly to something like an order statistic. Unfortunately, I haven't received funding to work on this, so I haven't published that, but I can say that an analytical solution likely exists to permit more robust FWER or FDR corrections of the P-values (Unfortunately, you can't just use p.adjust with n equal to the number of edges at that stage - that ends up being too conservative due to the dependence among edges).

However, if you wanted to justify that you're doing a conservative, multiple-hypothesis-testings/FWER correction, you can use p.adjust with n= (2*(m-factor+1)-3) in a p.adjust or similar step (this is approximating the number of hypotheses tested at each step. This becomes super conservative not just because the edges at any given step produce correlated results, but also because downstream steps' P-values tend to be bounded from below- they p-values aren't quite "the minimum of 2*(m-factor+1)-3 uniform random variables" because they tend to be greater than the minimum found at the previous step (sometimes, factoring out one edge can amplify the significance of downstream steps, hence these P-values aren't properly bounded, but their distribution is certainly shifted away from uniform if the null hypothesis is true).

Lots of open questions! There are ways to get more firm statistical statements in the event you're working with any clinical diagnostics or other studies that depend heavily on a firm knowledge of our certainty. The ways above are extra conservative approaches to whittle down what you're looking at to a minimal set of factors you can be highly confident in. Alternatively, one can use this as an exploratory tool (most inferential tools start off their lives as exploratory tools, such as PCA prior to the calculation of the Marcenko-Pastur distribution for the singular values of random matrices).

Enjoy!

**Comment by davidwrandall** [comment]
Thanks, as previously, for such a helpful reply - it really is very much appreciated.

I do hope you are able to get the funding you need to take this further. I will certainly cite your paper if the project I am working on ever gets published. Do let me know if there is anything else I can do to ensure you get the recognition you deserve for your help, which goes well beyond the call of duty!


David
________________________________

**Comment by reptalex** [comment]
Happy to help! Science is fun and phylofactorization has been a cool way to explore some exciting math & biology problems. Good luck with your paper, and I hope you find some cool lineages with this tool!

-Alex
